### PR TITLE
enhancement(networking) Allow configuration of http connection timeout

### PR DESCRIPTION
**Pull request by ChinYing-Li**
The feature itself is seemingly easy to implement, but I am having troubles making the newly added test case work. In particular, an assertion is needed:
```rust
assert_eq!(err, HealthcheckError::UnexpectedStatus{ status: http::StatusCode::REQUEST_TIMEOUT });
```
I tried to obtain `err` in several way but can't properly cast to `HealthcheckError`.

```rust 
let res = healthcheck.await(); // res is of type Result<(), Err>
let err = healthcheck.await().unwrap_err();  // err is of type Error
```
But err now wraps around a trait object `dyn StdError + std::marker::Send + std::marker::Sync`, and I could not find a way to cast it to`HealthcheckError`. Any pointers would be appreciated!

### CODE REVIEW

**Comment by bruceg** [comment]
> But err now wraps around a trait object `dyn StdError + std::marker::Send + std::marker::Sync`, and I could not find a way to cast it to`HealthcheckError`. Any pointers would be appreciated!

You can downcast from such an object using `downcast_ref<T>()`. In your case, this would look something like:
```rust
assert_eq!(err.downcast_ref::<HealthcheckError>(), HealthcheckError::UnexpectedStatus { status: http::StatusCode::REQUEST_TIMEOUT });
```
You can also use `matches!` when the error can't be constructed due to private fields:
```rust
assert!(matches!(err.downcast_ref::<HealthcheckError>(), HealthcheckError::UnexpectedStatus { ... }));
```

File: src/sinks/elasticsearch/config.rs [hunk 1]
```diff
@@ -307,7 +309,12 @@ impl SinkConfig for ElasticSearchConfig {
     async fn build(&self, cx: SinkContext) -> crate::Result<(VectorSink, Healthcheck)> {
         let common = ElasticSearchCommon::parse_config(self)?;
 
-        let http_client = HttpClient::new(common.tls_settings.clone(), cx.proxy())?;
+        let connection_timeout_duration = Duration::new(self.connection_timeout.unwrap_or(5), 0);
+        let http_client = HttpClient::new(
+            common.tls_settings.clone(),
+            cx.proxy(),
+            Some(connection_timeout_duration),
+        )?;
```
Comment (bruceg) [inline src/sinks/elasticsearch/config.rs:317]: Is this unwrap and then `Some` needed? The default timeout is 5 in the core `http.rs` implementation.

File: src/http.rs [hunk 2]
```diff
@@ -53,17 +54,25 @@ where
...
 
     pub fn new_with_custom_client(
         tls_settings: impl Into<MaybeTlsSettings>,
         proxy_config: &ProxyConfig,
+        timeout_duration: Option<Duration>,
         client_builder: &mut client::Builder,
     ) -> Result<HttpClient<B>, HttpError> {
         let mut http = HttpConnector::new();
         http.enforce_http(false);
+        http.set_connect_timeout(timeout_duration.or_else(|| Some(Duration::from_secs(5))));
```
Comment (bruceg) [inline src/http.rs:75]: I'd prefer this number be given a `const` near the top of this file to make it a little more explicit and easier to recognize if it needs changing.

**Comment by bruceg** [comment]
Thanks for this contribution, @ChinYing-Li. The code does look pretty good and straightforward as you say. I had a question and suggestion left inline above, and this will need merging with the current master as there are merge conflicts now.

**Comment by ChinYing-Li** [comment]
Thank you for the advice, @bruceg ! I just updated the test case to use a mock server, but still trying to achieve the desired connection timeout. I will `make fmt` the code again once done.

**Comment by jszwedko** [comment]
Closing this as stale, but feel free to reopen if this work is picked back up!

**Pull request closed without merging**